### DISCUSSION TITLE
​Mass Export of Workspace

### DISCUSSION THREAD

**Discussion by ntrkd** [toeverything/AFFiNE#6409] [Ideas]
### Description

When in a workspace on the homepage there should be a button that goes through and exports all the files to the selected export type.

### Use case

For backing up their workspace to a more universal format other than just the default affine file type or for converting their workspace to some other format.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by JimmFly** [comment]
What do you think about this? @hwangdev97

**Comment by yinan-c** [comment]
I like the export functionality that converts a single page to a markdown file, it would be good to be able to export the whole workspace.

**Comment by IvanLi-CN** [comment]
I want this feature too, it would give me a very big sense of security.

**Comment by deltaxsingh** [comment]
Seems like Affine could easily implement this as the functionality to "export as markdown" already exists per page. 

I'm guessing they don't want users leaving so it's intentionally this way. Especially since the original request was from March 2024. 

Pretty scummy.

**Reply by louis-lau** [reply]
Are you not a developer, or perhaps a junior developer? Often, more goes into such a task than you think. You can't know if they left this intentionally, they're probably just not prioritizing it. Calling people scummy for not implementing feature requests in an open source project is the kind of thing that makes open source developers burn out and quit.

**Reply by deltaxsingh** [reply]
> Are you not a developer, or perhaps a junior developer? Often, more goes into such a task than you think. You can't know if they left this intentionally, they're probably just not prioritizing it. Calling people scummy for not implementing feature requests in an open source project is the kind of thing that makes open source developers burn out and quit.

Easily solution is to not start making a product scummy to begin with mate. In no world is import without adequate export reasonable, irrespective of "priorities".

**Comment by rrmistry** [comment]
Found this JS script that seems to perform a basic import/export of all Affine data
> https://gist.github.com/pengx17/6e51885d892b0ce2a29e5f5c0e2d84fe

**⚠️ Disclaimer:** I haven't personally tested and I am _**not**_ the author

**Reply by marcolaux** [reply]
this is not a universal format and the "full backup" feature within affine should do the same I think.

**Comment by marcolaux** [comment]
This would be an important feature for portability and accessibility

**Comment by olig89** [comment]
I was surprised to not find this as an option, especially with there being a "Select All" option on the All Docs screen, only used for deletion.

I _would_ use the "full backup" feature but that requires a full sync before it can export and currently my self hosted server is down - the prime use case for bulk backing up what I have on my local machine.

**Comment by CMCDragonkai** [comment]
I'd like to have this because sometimes I need to do a full dump of a workspace into an LLM for further analysis... but right now the snapshot JSON appears to be something custom? Like in some cases, only content is needed, no need for all the formatting and styling metadata. Text in a way is the universal interface... graphical relations - can maybe always be transferred with some information loss into text.

**Comment by mobamoba** [comment]
I'd like to have this because sometimes developers abandon projects for various reasons, and then my data is simply will simply be stuck in Affine and unusable.

**Comment by lorrod** [comment]
Custom integrations are essential for my setup — without them, it becomes impossible to use effectively.

**Comment by amunrarara** [comment]
I'm hoping to, at least, perform a fully-nested export similar to the Notion feature I used to import _into_ Affine. It seems exporting gives me a single flat file without nested items, which is quite frustrating for moving files between Affine workspaces. 

Plus, if I export a page, all the internal page links still point to the _old_ workspace.

**Reply by amunrarara** [reply]
Also, if I wanted to move off AFFiNE Cloud and onto a local install, it doesn't seem that's an easy transition to make. There's a project from a community member that I haven't tested, https://github.com/pengx17/affine-reader and also https://gist.github.com/pengx17/6e51885d892b0ce2a29e5f5c0e2d84fe from the same user, but that should be a core feature, yeah?

**Comment by saaqb** [comment]
There is no full workspace all in one file export functionality except the .affine format.

**Comment by Danie10** [comment]
It was only because Joplin Notes had a mass export to Markdown, that I am now here at AFFiNE with my 1,000+ notes. But it is not to say I'll be using AFFiNE forever, and AFFiNE may also not be around forever. There should always be some way to export data into an interoperable format. I understand that, just like Joplin Notes, there may be a more efficient way of processing the working data, but there should still be an easy way to export years of one's work, otherwise it is locked into the ecosystem. Manually exporting each one of 1000+ notes to Markdown format is not very practical. I prefer my backups also to be in plain text MD format.

**Comment by Danie10** [comment]
I was going to suggest logging an issue for this, but I see there was one logged as #6353, and it was closed. 

I've really enjoyed using AFFiNE in the one week I've been using, and I have been preparing this last week to feature it in a video I want to do about it, but this is a bit of a showstopper for me.  I either don't do the video or I must make a massive caveat warning about the lock-in. I also did not realise this until after I'd imported just under 2,000 notes from my Joplin Notes hosting.

It would be useful to at least hear from AFFiNE whether there will in future be a bulk export to Markdown or not.

I'm probably going to do the video anyway as I have spent so much time on it already.

**Reply by Anazark** [reply]
Please do your video and mention this caveat so that others don't fall in this trap

**Comment by Anazark** [comment]
I too am in this situation where the import was easy, the backup too, but the export back to MD in batch is nowhere to be found
And I think the same as deltaxsingh : not having developed the functionnality is one thing, but the lack of transparency and lack of answers to this core issue is indeed problematic
To the Affine team, you're doing a great job, but please this is an urgent matter

**Reply by qredwoods** [reply]
Same issue. I found Affine elegant and delightful, but blocked by fear of lock-in / data loss. For alternatives meanwhile, I'm now looking at AnyType which has the mass export to MD feature. (different paradigm but similar UI post-Notion privacy-oriented tool, open-sourcing since 2023, some still to go, can't toggle off telemetry but that's planned, your files stay home, encrypted. Obsidian doesn't work for my needs/wants.) From below sounds like it's close though. Good luck all!

**Comment by Danie10** [comment]
I've been looking at the existing single document export to MD, and it has interesting possibilities:

1. If instead of naming the created .zip file with the name of the document, it could be changed to affineexport.zip or similar.
2. The created index.md file could instead be named as the name of the individual document.md and in this way many could be added to the same .zip export.
3. The assets sub-folder can stay as is for all attachments and media.
4. The assets already exported can also stay as is, inside the assets sub-folder.
5. The media links inside each document though, should then point to /assets/media.jpg or whatever. Right now this is not even happening for a single document, so the assets get exported, but the links are not updated, thereby breaking the links.

It is sort of nearly there, and with a few tweaks, this could be a great bulk export tool.

**Comment by DanielBCoffaro** [comment]
Please add the ability to export a workspace. Preferably as markdown. Without it I am hesitant to go all in as my notes are a bit trapped. 
Thank you for the work you have done on Affine. It's a great application and I want to use it without reservation.

**Comment by CMCDragonkai** [comment]
It's interesting that the Affine desktop application is electron, and its local workspace is just wherever the electron app is storing the data. The documented way to store local data is to run the affine server and configure its storage. However ideally, the local workspace can also be specified instead of just sticking it wherever electron puts its local data. Then the local data is probably some sort of SQL database. In that case, you'd need something to sync the SQL database. And ideally something that bridges the gap between text file syncing to db level syncing. There are some bridges but they aren't the best. It reminds me of DAT.

**Reply by tkodev** [reply]
Two things: 

- Local storage: Affine can run locally, or in self hosted domain stored mode (press login to self hosted instance). It comes out of the box with a local instance (for easy new user onboarding I assume)

- DB: A great majority of the self hosted apps on the internet store in some kind of DB that can be read. DB isn't the issue, the issue is having to reverse engineer and come out with third party tools just so we can do this. Not to mention the maintenance for such a tool / out of sync db migrations, and whether affine will one day target them (given that it will be a source of customer complaints, and that they already don't like exporting in a common format). Perhaps it's a simple lack of resourcing or not. Unsure of affine's stance

Having build block based headless CMSes before (notion/affine are basiclly this but with a desktop app), complexity of reverse engineering really depends on how they modelled the db (object / block based)

**Comment by tkodev** [comment]
This is a serious issue for me. Why is mass exporting to a common format like HTML (used in notion -> affine import) or markdown (used in simple non-block apps) not an option? It's such an proprietary move for a self hosted / open source application.

**Comment by Danie10** [comment]
@tkodev for me too, which is why I had to just move on to Anytpe so long. I was hoping that AFFiNE would at least give a solid commitment, but without that, it was not worth me just waiting in hope.

**Comment by florentineprinzessinzusachsen** [comment]
This tampermonkey script might help some people who want to move on from Affine. I switched to anytype as well. Didn't want to put all my data into a "closed" system with limited export functionality. Hope devs will consider bulk export for the future.

Tested this on 0.26.0, might break in the future. Made it partly using claude so super flaky but working.

https://gist.github.com/florentineprinzessinzusachsen/c7b17222d3910c6c412f783ed25377ce

**Reply by Anazark** [reply]
THANK
YOU
SO
MUUUUUCH

It worked almost perfectly, today is the day that you freed me from the shackles of Affine, you're the goat

I would just add a few things that weren't obvious for me : 
1) You need to open your docs on the website "https://app.affine.pro" and add "https://app.affine.pro/*" to the "@match" line in the script
2) Then when you reload the affine page, some buttons will appear in the upper right corner
3) BEFORE you launch the export, go to your browser settings and uncheck "always ask where to download files" so that all files are automatically downloaded without your input for each one of them
4) Then launch the export and go watch something in the background or do whatever
5) Finally, check if the correct number of files was output (for me, 13 out of 140 weren't so I exported those manually)

**Reply by nkrohit13** [reply]
tried implementing this, isn't working

**Reply by Anazark** [reply]
Well you're gonna need to give us more info if you want us to be able to help
What exactly isn't working, at which step, what error do you get, what browser did you use, etc

Hope you'll be able to find out a solution

**Reply by nkrohit13** [reply]
Hi, 
Sorry for this, my bad. 
I forgot to enable Allow User Scripts options in Tampermonkey settings and after restarting the browser it worked. 
Also enable multiple download option from site setting beforehand.
Working like a charm.
:))

**Reply by TheSpyder** [reply]
Thank you for this, it helped me a lot (there haven’t been any releases since February so there’s no concern about compatibility.

Luckily I only have 26 docs, I can imagine bigger instances finding it tough to deal with a flat export that doesn’t keep the folder structure.

**Reply by martin861101** [reply]
I enabled MCP and an agent exported the full workspace within 2 minutes.

**Method:**

1. Enable MCP in settings
2. Copy provides JSON string (Bearer token is included)
3. Add MCP config to your agent config.json file
4. Ask the agent to export all documents in the specified directory.

I used opencode with a free agent.

Directory for MCP: ~/.config/opencode/opencode.jsonc

Your conf should look like this:
```
{
  "$schema": "https://opencode.ai/config.json",
  "mcp": {
    "affine_workspace_name": {
      "type": "remote",
      "url": "http(s)://{AFFiNE_URL}/api/workspaces/{ID_String}/mcp",
      "enabled": true,
      "oauth": false,
      "headers": {
        "Authorization": "Bearer YOUR_TOKEN"
      }
    }
  }
}
